### PR TITLE
WheelPicker - generic value

### PR DESCRIPTION
**Pull request by Inbal-Tish**
## Description
WheelPicker - generic value

## Changelog
WheelPicker - generic value

## Additional info
ticket 1844

### CODE REVIEW

**Comment by Inbal-Tish** [comment]
@M-i-k-e-l Ping

File: src/components/WheelPicker/index.tsx [hunk 1]
```diff
@@ -97,30 +97,34 @@ export interface WheelPickerProps {
...
-  disableRTL,
-  style,
-  children,
-  initialValue = 0,
-  separatorsStyle,
-  testID,
-  faderProps,
-  flatListProps
-}: WheelPickerProps) => {
+const WheelPicker = <T extends string | number = number>(props: WheelPickerProps<T>) => {
```
Comment (M-i-k-e-l) [inline src/components/WheelPicker/index.tsx:103]: I think we can remove the default here: `<T extends string | number>`
Also, we have this type here as well

File: src/components/WheelPicker/usePresenter.ts [hunk 2]
```diff
@@ -1,59 +1,57 @@
 import _ from 'lodash';
 import React from 'react';
-import {ItemProps} from './Item';
+import {LogService} from '../../services';
 import useMiddleIndex from './helpers/useListMiddleIndex';
+import {ItemProps} from './Item';
 
-export type ItemValueTypes = ItemProps | number | string;
+//TODO: deprecate this type
+export type ItemValueTypes = /* ItemProps |  */number | string;
```
Comment (M-i-k-e-l) [inline src/components/WheelPicker/usePresenter.ts:8]: We have this type here as well

File: src/components/WheelPicker/usePresenter.ts [hunk 3]
```diff
@@ -1,59 +1,57 @@
...
-interface Presenter {
-  items: ItemProps[];
-  index: number;
-  height: number;
-  getRowItemAtOffset: (offset: number) => RowItem;
-}
-
-const usePresenter = ({
-  initialValue = 0,
+const usePresenter = <T extends string | number = number>({
```
Comment (M-i-k-e-l) [inline src/components/WheelPicker/usePresenter.ts:23]: We have this type here as well
Don't think we need the default `= number`

File: src/components/WheelPicker/usePresenter.ts [hunk 4]
```diff
@@ -1,59 +1,57 @@
...
   };
-
   const items = children ? extractItemsFromChildren() : propItems || [];
   const middleIndex = useMiddleIndex({itemHeight, listSize: items.length});
 
   const getSelectedValueIndex = () => {
-    if (_.isString(initialValue) || _.isNumber(initialValue)) {
-      return _.findIndex(items, {value: initialValue});
+    if (_.isObject(initialValue)) {
+      LogService.warn('UILib WheelPicker will stop supporting initialValue prop type as an object (ItemProps). Please pass string or number only');
```
Comment (M-i-k-e-l) [inline src/components/WheelPicker/usePresenter.ts:43]: I'm assuming this is what this refers to: `export type ItemValueTypes = /* ItemProps |  */number | string;`
If we're only starting this migration we should probably only remove the type in the next major when we remove it.

File: src/components/sectionsWheelPicker/SectionsWheelPicker.driver.tsx [hunk 5]
```diff
@@ -3,9 +3,9 @@ import {useComponentDriver, ComponentProps} from '../../testkit/new/Component.dr
 import {WheelPickerDriver} from '../WheelPicker/WheelPicker.driver';
 import {SectionsWheelPickerProps} from './index';
 
-export const SectionsWheelPickerDriver = (props: ComponentProps) => {
-  const driver = useComponentDriver<SectionsWheelPickerProps>(props);
-  const sections = driver.getProps().children as SectionsWheelPickerProps;
+export const SectionsWheelPickerDriver = <T extends string | number = number>(props: ComponentProps) => {
```
Comment (M-i-k-e-l) [inline src/components/sectionsWheelPicker/SectionsWheelPicker.driver.tsx:6]: We have this type here as well
Don't think we need the default = number

File: src/components/sectionsWheelPicker/index.tsx [hunk 6]
```diff
@@ -42,9 +43,18 @@ export type SectionsWheelPickerProps = PropsWithChildren<{
  * @gif: https://github.com/wix/react-native-ui-lib/blob/master/demo/showcase/SectionsWheelPicker/SectionsWheelPicker.gif?raw=true
  */
 
-const SectionsWheelPicker = (props: SectionsWheelPickerProps) => {
-  const {sections, itemHeight, numberOfVisibleRows, activeTextColor, inactiveTextColor, textStyle, disableRTL, testID} =
-    props;
+const SectionsWheelPicker = <T extends string | number = number>(props: SectionsWheelPickerProps<T>) => {
```
Comment (M-i-k-e-l) [inline src/components/sectionsWheelPicker/index.tsx:46]: We have this type here as well
Don't think we need the default = number

**Review by M-i-k-e-l — COMMENTED** [review]
Nice work, mostly recurring comments, I did not want to miss them if we make changes

File: src/components/WheelPicker/usePresenter.ts (hunk 4 above)
Comment (Inbal-Tish) [reply src/components/WheelPicker/usePresenter.ts:43]: We still support it so functionality won't break but they will get TS error that should force them to migrate or add ignore. WDYT?

File: src/components/WheelPicker/Item.tsx [hunk 7]
```diff
@@ -1,23 +1,23 @@
...
+import {useThemeProps} from '../../hooks';
+import Text, {TextProps} from '../text';
+import TouchableOpacity from '../touchableOpacity';
 import {WheelPickerAlign} from './types';
 
 const AnimatedTouchableOpacity = Animated.createAnimatedComponent(TouchableOpacity);
 const AnimatedText = Animated.createAnimatedComponent(Text);
 
-export interface ItemProps {
+export interface ItemProps<T> {
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/WheelPicker/Item.tsx:13]: I'm not sure how to pass the type from the Item either, I'll take a look.
Regarding the DefaultWheelPickerItemType we can use the existing `ItemValueTypes`, no?

File: src/components/WheelPicker/Item.tsx [hunk 8]
```diff
@@ -99,9 +102,9 @@ const WheelPickerItem = memo(({
       )}
     </AnimatedTouchableOpacity>
   );
-});
+};
 
-export default asBaseComponent<InternalProps>(WheelPickerItem);
+export default memo(WheelPickerItem) as typeof WheelPickerItem;
```
[the remark this answers is not in the document]
Comment (Inbal-Tish) [reply src/components/WheelPicker/Item.tsx:107]: Don't remember... removed

File: src/components/WheelPicker/Item.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/WheelPicker/Item.tsx]: I guess so (missed that), but not sure about the name, I feel like it should have `default` in it

File: src/components/WheelPicker/usePresenter.ts (hunk 2 above)
Comment (M-i-k-e-l) [reply src/components/WheelPicker/usePresenter.ts:8]: Generally yes, lets move this to the above thread

File: src/components/WheelPicker/usePresenter.ts (hunk 4 above)
Comment (M-i-k-e-l) [reply src/components/WheelPicker/usePresenter.ts:43]: We don't usually do that IMO, we don't cause TS errors if we can avoid it

File: src/components/sectionsWheelPicker/SectionsWheelPicker.driver.tsx (hunk 5 above)
Comment (M-i-k-e-l) [reply src/components/sectionsWheelPicker/SectionsWheelPicker.driver.tsx:6]: Still have the default

File: src/components/WheelPicker/usePresenter.ts (hunk 4 above)
Comment (Inbal-Tish) [reply src/components/WheelPicker/usePresenter.ts:43]: I remane it WheelPickerItemValue and defined it in types so I can export it from WheelPicker. I also exported outside

**Review by M-i-k-e-l — APPROVED** [review]

**Pull request merged**